include/Imap/ImapHandler.php

```diff
@@ -717,16 +717,26 @@ public function rfc822ParseHeaders($headers, $defaulthost = "UNKNOWN")
     public function search($criteria, $options = SE_FREE, $charset = null)
     {
         // Default to class charset if none is specified
-        $emailCharset = (!empty($charset)) ? $charset : $this->charset;
+        $emailCharset = !empty($charset) ? $charset : $this->charset;
 
         $this->logCall(__FUNCTION__, func_get_args());
 
-        $ret = imap_search($this->getStream(), $criteria, $options, $emailCharset);
+        try {
+            $ret = imap_search($this->getStream(), $criteria, $options, $emailCharset);
+        } catch (Exception $e) {
+            if (strpos($e, ' [BADCHARSET (US-ASCII)]')) {
+                LoggerManager::getLogger()->debug("Encoding changed dynamically from {$emailCharset} to US-ASCII");
+
+                $emailCharset = 'US-ASCII';
+                $ret = imap_search($this->getStream(), $criteria, $options, $emailCharset);
```

Comment:
unhandled case, return value from imap_search could be:
"Return FALSE if it does not understand the search criteria or no messages have been found."

Reply:
@gymad There's a check for this on line 735